iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp

```diff
@@ -24,6 +24,62 @@
 namespace mlir {
 namespace iree_compiler {
 
+void ConvertToDynamicSharedMemory(ModuleOp moduleOp) {
+  // Collect all the adressOfOps to static shared memory globals.
+  SmallVector<LLVM::AddressOfOp> addressOfOps;
+  moduleOp.walk([&](LLVM::AddressOfOp addressOfOp) {
+    if (!addressOfOp.getGlobal().getConstant())
+      addressOfOps.push_back(addressOfOp);
+  });
+  if (addressOfOps.size() == 0) return;
+  OpBuilder builder(moduleOp);
+  builder.setInsertionPoint(&moduleOp.front());
+  auto type =
+      LLVM::LLVMArrayType::get(IntegerType::get(builder.getContext(), 8), 0);
+  LLVM::GlobalOp global = builder.create<LLVM::GlobalOp>(
+      moduleOp.getLoc(), type, /*isConstant=*/false, LLVM::Linkage::External,
+      "__dynamic_shared_memory__", Attribute(),
+      /*alignment=*/16, /*addr_space=*/3);
+  int numberOfBytes = 0;
+  // Replace the addressOfOps with correctly offseted pointers to dynamic
+  // shared memory
+  for (auto addressOfOpsIt : llvm::enumerate(addressOfOps)) {
+    auto addressOfOp = addressOfOpsIt.value();
+    auto loc = addressOfOp.getLoc();
+    builder.setInsertionPoint(addressOfOp);
+    auto globalOp = addressOfOp.getGlobal();
+    LLVM::AddressOfOp globalPtr =
+        builder.create<LLVM::AddressOfOp>(loc, global);
+    Value zero = builder.create<LLVM::ConstantOp>(
+        loc, IntegerType::get(builder.getContext(), 64),
+        builder.getI64IntegerAttr(0));
+    Value offset = builder.create<LLVM::ConstantOp>(
+        loc, IntegerType::get(builder.getContext(), 64),
+        builder.getI64IntegerAttr(numberOfBytes));
+    Value shiftedPtr = builder.create<LLVM::GEPOp>(
+        loc, globalPtr.getType(), globalPtr, ValueRange({zero, offset}));
+    Value castPtr = builder.create<LLVM::BitcastOp>(
+        loc,
+        LLVM::LLVMPointerType::get(globalOp.getType(), global.getAddrSpace()),
+        shiftedPtr);
+    addressOfOp.replaceAllUsesWith(castPtr);
+    auto thisarray = globalOp.getType();
+    // calculate offset for next allocation
+    DataLayout dataLayout = DataLayout::closest(addressOfOp);
+    numberOfBytes += dataLayout.getTypeSizeInBits(thisarray) / 8;
```

Comment:
This still assume there is a single addressOfOp per global. I think to avoid any problem we either need to only consider globalOp with a single use when we collect the ops or we need to loop though the global instead.

Reply:
Makes sense, I think the later is easier with the hasoneuse function

Reply:
ok, it should be easy to change your code to collect globals instead and have the same logic. The get the addressOfOp you can just loop through the use with something like:
```
for(OpOperand& use : globalOp.getUses()) {
...
}
```